### PR TITLE
fix(amazonq): show more specific error messages

### PR DESCRIPTION
**Pull request by dhasani23**
## Problem

Need to show the user more specific error messages for various failure cases during a transformation.

## Solution

Handle specific errors with specific messaging, text already approved by our technical writer.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts [hunk 1]
```diff
@@ -187,10 +187,8 @@ export async function startTransformByQ() {
     } catch (error: any) {
         await transformationJobErrorHandler(error)
     } finally {
-        if (transformByQState.isCancelled()) {
-            await postTransformationJob()
-            await cleanupTransformationJob()
-        }
+        await postTransformationJob()
+        await cleanupTransformationJob()
```
Comment (dhasani23) [inline packages/core/src/codewhisperer/commands/startTransformByQ.ts:191]: These should always be running since they reset the Transformation Hub and other things to prepare for a subsequent transformation run.

File: packages/core/src/amazonqGumby/chat/controller/messenger/messenger.ts [hunk 2]
```diff
@@ -384,11 +384,7 @@ export class Messenger {
         this.dispatcher.sendCommandMessage(new SendCommandMessage(message.command, message.tabID, message.eventId))
     }
 
-    public sendJobFinishedMessage(tabID: string, message: string = '') {
-        if (message === '') {
```
Comment (dhasani23) [inline packages/core/src/amazonqGumby/chat/controller/messenger/messenger.ts:388]: This never happened.

File: packages/core/src/codewhisperer/service/transformByQ/transformApiHandler.ts [hunk 3]
```diff
@@ -178,16 +182,12 @@ export async function resumeTransformationJob(jobId: string, userActionStatus: T
...
+        throw new Error(errorMessage)
     }
 }
 
 export async function stopJob(jobId: string) {
     if (!jobId) {
-        throw new Error('Job ID is empty')
-    }
-
-    if (transformByQState.isNotStarted()) {
```
Comment (dhasani23) [inline packages/core/src/codewhisperer/service/transformByQ/transformApiHandler.ts:190]: Duplicate check; already handled where we call `stopJob`

File: packages/core/src/codewhisperer/service/transformByQ/transformApiHandler.ts [hunk 4]
```diff
@@ -178,16 +182,12 @@ export async function resumeTransformationJob(jobId: string, userActionStatus: T
...
             reason: 'ResumeTransformationFailed',
         })
-        throw new Error('Resume transformation job failed')
+        throw new Error(errorMessage)
     }
 }
 
 export async function stopJob(jobId: string) {
     if (!jobId) {
-        throw new Error('Job ID is empty')
```
Comment (dhasani23) [inline packages/core/src/codewhisperer/service/transformByQ/transformApiHandler.ts:187]: Don't throw an error since that gets surfaced to customer; just avoid calling `stopJob` if no `jobId` has been generated yet.

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts [hunk 5]
```diff
@@ -479,16 +484,28 @@ export async function startTransformationJob(uploadId: string) {
...
-        if ((error as Error).message.includes('too many active running jobs')) {
+        const errorMessage = (error as Error).message
+        if (errorMessage.includes('too many active running jobs')) {
             transformByQState.setJobFailureErrorNotification(
                 CodeWhispererConstants.failedToStartJobTooManyJobsNotification
             )
             transformByQState.setJobFailureErrorChatMessage(
                 CodeWhispererConstants.failedToStartJobTooManyJobsChatMessage
             )
+        } else if (errorMessage.includes('Lines of Code limit breached')) {
```
Comment (JuliaABurch) [inline packages/core/src/codewhisperer/commands/startTransformByQ.ts:495]: Would it be better to create custom Error types and switch on the type rather than looking at strings? If strings get updated in one place but not everywhere, then these lines of code break

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts [hunk 6]
```diff
@@ -479,16 +484,28 @@ export async function startTransformationJob(uploadId: string) {
         getLogger().info(`CodeTransformation: jobId: ${jobId}`)
     } catch (error) {
         getLogger().error(`CodeTransformation: ${CodeWhispererConstants.failedToStartJobNotification}`, error)
-        if ((error as Error).message.includes('too many active running jobs')) {
+        const errorMessage = (error as Error).message
+        if (errorMessage.includes('too many active running jobs')) {
```
Comment (JuliaABurch) [inline packages/core/src/codewhisperer/commands/startTransformByQ.ts:488]: It would be nice if we could centralize all of this error handling (setting the notification / chat messages) in one class, as right now this functionality gets scattered across several different files / classes

File: packages/core/src/test/codewhisperer/commands/transformByQ.test.ts [hunk 7]
```diff
@@ -131,25 +131,18 @@ describe('transformByQ', function () {
...
     })
 
     it('WHEN stop job called with valid jobId THEN stop API called', async function () {
-        transformByQState.setToRunning()
         const stopJobStub = sinon.stub(codeWhisperer.codeWhispererClient, 'codeModernizerStopCodeTransformation')
         await stopJob('dummyId')
         sinon.assert.calledWithExactly(stopJobStub, { transformationJobId: 'dummyId' })
     })
 
-    it('WHEN stop job that has not been started THEN stop API not called', async function () {
```
Comment (JuliaABurch) [inline packages/core/src/test/codewhisperer/commands/transformByQ.test.ts:147]: Do we not need this edge case / is it tested elsewhere?

Comment (dhasani23) [reply packages/core/src/test/codewhisperer/commands/transformByQ.test.ts:147]: No longer needed since in `stopJob` I removed a check that checks if the job is marked as "running" or not, since that was a redundant check as it's handled elsewhere.

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts (hunk 5 above)
Comment (dhasani23) [reply packages/core/src/codewhisperer/commands/startTransformByQ.ts:495]: We'd still have to check the Strings though right? This String comes directly from backend.

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts (hunk 6 above)
Comment (dhasani23) [reply packages/core/src/codewhisperer/commands/startTransformByQ.ts:488]: Yeah agreed, from UX we soon will have to modify this so that error messages only show in notification if the chat is not visible to user, when that work is done I think will be a good time to refactor how we set the error messages.

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts (hunk 1 above)
Comment (tincheng) [reply packages/core/src/codewhisperer/commands/startTransformByQ.ts:192]: finalizeTransformByQ (down the chain of humanInTheLoopRetryLogic for happy path) will make duplicated call to postTransformationJob and cleanupTransformationJob -> can this cause extra messages to be published to chat or extra telemetry to be emitted? 

```
export async function finalizeTransformByQ(status: string) {
    try {
        // Set the result state variables for our store and the UI
        await finalizeTransformationJob(status)
    } catch (error: any) {
        await transformationJobErrorHandler(error)
    } finally {
        await postTransformationJob()
        await cleanupTransformationJob()
    }
}
```
I think we can remove the postTransformationJob() and cleanupTransformationJob() from finalizeTransformByQ

File: packages/core/src/codewhisperer/models/constants.ts [hunk 8]
```diff
@@ -451,37 +454,47 @@ export const unsupportedJavaVersionChatMessage =
...
 
 export const failedToStartJobTooManyJobsChatMessage =
     'Sorry, I couldn’t begin the transformation. You have too many active transformations running. Please try again after your other transformations have completed.'
 
 export const failedToStartJobTooManyJobsNotification =
-    "Amazon Q couldn't begin the transformation. You have too many active transformations running. Please try again after your other transformations have completed."
+    'Amazon Q could not begin the transformation. You have too many active transformations running. Please try again after your other transformations have completed.'
+
+export const failedToStartJobLinesLimitNotification =
+    'Amazon Q could not begin the transformation. Your code exceeds the lines of code limit. Please try again with a smaller project.'
```
Comment (tincheng) [inline packages/core/src/codewhisperer/models/constants.ts:469]: There are currently two scenarios for failedToStartJobLinesLimitNotification:
1. "Job Lines of Code limit breached" - single project exceeded line limit.
2. "Monthly aggregated Lines of Code limit breached" - completed jobs in the current month exceeded aggregated line limit across jobs.

This message always assume case 1.

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts (hunk 6 above)
Comment (justinmk3) [reply packages/core/src/codewhisperer/commands/startTransformByQ.ts:488]: there are some useful patterns (and util functions) in `packages/core/src/shared/errors.ts`

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts (hunk 5 above)
Comment (damntrecky) [reply packages/core/src/codewhisperer/commands/startTransformByQ.ts:495]: Not a blocker: I think @JuliaABurch is asking about semantic readability and maybe a small separation of concerns. The code might read better for developers if it was something like

`else if (error === LinesOfCodeLimitBreachedError)` and you abstract the "logic" of the string checking in the error types we create.

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts [hunk 9]
```diff
@@ -479,16 +481,35 @@ export async function startTransformationJob(uploadId: string) {
...
             transformByQState.setJobFailureErrorChatMessage(
                 CodeWhispererConstants.failedToStartJobTooManyJobsChatMessage
             )
+        } else if (errorMessage.includes('Monthly aggregated Lines of Code limit breached')) {
+            transformByQState.setJobFailureErrorNotification(
+                CodeWhispererConstants.failedToStartJobMonthlyLimitNotification
+            )
+            transformByQState.setJobFailureErrorChatMessage(
+                CodeWhispererConstants.failedToStartJobMonthlyLimitChatMessage
+            )
```
Comment (tincheng) [inline packages/core/src/codewhisperer/commands/startTransformByQ.ts:498]: Sorry didn't catch this last time, but Lines of Code limit is checked during the transformation build phase and will be returned as a transformation failure reason, instead of a validation error from start job. These should be handled during transformation polling.
